napari/layers/transforms.py

```diff
@@ -0,0 +1,173 @@
+        self.func = func
+        self._inverse_func = inverse
+        self.name = name
+
+        if func is tz.identity:
+            self._inverse_func = tz.identity
+
+    def __call__(self, coords):
+        """Transform input coordinates to output."""
+        return self.func(coords)
+
+    def set_slice(self, axes: Sequence[int]) -> Transform:
+        """Return a transform subset to the visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    def set_pad(self, axes: Sequence[int]) -> Transform:
+        """Return a transform with added axes for non-visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    @property
+    def inverse(self):
+        if self._inverse_func is not None:
+            return Transform(self._inverse_func, self.func)
+        else:
+            raise ValueError('Inverse function was not provided.')
+
+
+# class TransformChain(ListModel, Transform):
+#     def __init__(self, transforms=[]):
+#         super().__init__(
+#             basetype=Transform,
+#             iterable=transforms,
+#             lookup={str: lambda q, e: q == e.name},
+#         )
+#
+#     def __call__(self, coords):
+#         return tz.pipe(coords, *self)
+#
+#     def __newlike__(self, iterable):
+#         return ListModel(self._basetype, iterable, self._lookup)
+#
+#     def set_slice(self, axes: Sequence[int]) -> TransformChain:
+#         return TransformChain([tf.set_slice(axes) for tf in self])
+
+
+class STTransform(Transform):
+    """n-dimensional scale and translation (shift) class.
+
+    Scaling is always applied before translation.
+
+    An empty translation vector implies no translation, and an empty scale
+    implies no scaling.
+
+    Translation is broadcast to 0 in leading dimensions, so that, for example,
+    a translation of [4, 18, 34] in 3D can be used as a translation of
+    [0, 4, 18, 34] in 4D without modification.
+
+    Translation is broadcast to 1 in leading dimensions, so that, for example,
+    a scale of [4, 18, 34] in 3D can be used as a scale of
+    [1, 4, 18, 34] in 4D without modification.
+    """
+
+    def __init__(self, scale=(1.0,), translate=(0.0,), name=None):
+        super().__init__(name=name)
+        self.scale = np.array(scale)
+        self.translate = np.array(translate)
+
+    def __call__(self, coords):
+        coords = np.atleast_2d(coords)
+        scale = np.concatenate(
+            ([1.0] * (coords.shape[1] - len(self.scale)), self.scale)
+        )
+        translate = np.concatenate(
+            ([0.0] * (coords.shape[1] - len(self.translate)), self.translate)
+        )
+        return np.squeeze(scale * coords + translate)
+
+    @property
+    def inverse(self):
+        return STTransform(1 / self.scale, -self.translate)
+
+    def set_slice(self, axes: Sequence[int]) -> STTransform:
+        return STTransform(self.scale[axes], self.translate[axes])
```

Comment:
Might this have issues when the `scale` is set at one dimensionality, but the viewer has since increased in ndims, and now axes is too big for the original `self.scale`?

Reply:
The code here is meant to prevent that https://github.com/napari/napari/pull/1018/files#diff-9ef2d7d09674e6ab30a1456a11fd74a8R350-R359